libvast/src/command.cpp

```diff
@@ -0,0 +1,258 @@
+    auto value = to<data>(*y);
+    if (!value)
+      return ec::syntax_error;
+    i->second.value = std::move(*value);
+    return {};
+  };
+  auto parse_long_opt = [&](auto& cmd, auto x) -> error {
+    if (!starts_with(*x, "--"))
+      return ec::syntax_error;
+    auto o = x.substr(2);
+    auto pred = [](auto& name, auto&) { return name == o; };
+    auto i = std::find_if(cmd.options.begin(), cmd.options.end(),pred);
+    if (i == cmd.options.end())
+      return ec::syntax_error;
+    auto s = split_to_str(o, "=");
+    if (is_toggle(i->second)) {
+      if (s.size() > 1)
+        return ec::syntax_error;
+      i->second.value = true;
+    } else {
+      if (s.size() == 1)
+        return ec::syntax_error;
+      if (auto value = to<data>(s[1]))
+        i->second.value = std::move(*value);
+      else
+        return ec::syntax_error;
+    }
+    return {};
+  };
+  auto parse_flag_or_long_opt = [&](auto& cmd, auto x) -> error {
+    return parse_short_opt(cmd, x) ? parse_long_opt(cmd, x) : error{};
+  };
+  auto parse_command = [&](auto& cmd, auto x) -> error {
+    auto pred = [&](auto& name, auto&) { return name == *x; };
+    auto i = std::find_if(cmd.commands.begin(), cmd.commands.end(), pred);
+    if (i == cmd.commands.end())
+      return ec::syntax_error;
+    return {};
+  };
+  auto copy = make(description);
+  copy->options = options;
+  copy->commands = commands;
+  auto i = args.begin();
+  auto j = i;
+  auto end = args.end();
+  while (i != end) {
+    if (++j == end) {
+      // With only one element left, we must have a short option or a
+      // command.
+      auto e = parse_flag_or_long_opt(*copy, *i);
+      if (!e)
+        return; // We're done, a new command starts here.
+      e = parse_command(*i);
+      if (e == ec::parse_error)
+        return {i, e};
+      return {++i, error{}};
+    }
+    // When dealing with a pair (i, j), we must have a short option with
+    // value or a command (i) with one argument (j).
+    auto e = parse_short_opt_with_value(i, j);
+    if (!e) {
+      ++i;
+      ++j;
+    }
+    if (e == ec::syntax_error || e != ec::parse_error)
+      return {i, e};
+    e = parse_command(*i);
+    if (e == ec::syntax_error || e != ec::parse_error)
+      return {i, e};
+    return commands[*i]->parse(++i, end); // recurse
+  }
+}
+*/
+
+std::string command::full_name() {
+  std::string result;
+  if (is_root())
+    return result;
+  result = parent_->full_name();
+  if (!result.empty())
+    result += ' ';
+  result += name_;
+  return result;
+}
+
+std::string command::name() {
+  return std::string{name_.begin(), name_.end()};
+}
+
+bool command::is_root() const noexcept {
+  return parent_ == nullptr;
+}
+
+command::proceed_result command::proceed(caf::actor_system&, opt_map& options,
+                                         caf::message args) {
+  CAF_LOG_TRACE(CAF_ARG(options) << CAF_ARG(args));
+  CAF_IGNORE_UNUSED(options);
+  CAF_IGNORE_UNUSED(args);
```

Comment:
`VAST_TRACE` can't print `caf::message´. Once that's fixed we can switch back to the VAST macros.

Reply:
Wouldn't `to_string(msg)` work with VAST's logging macros? I would imagine that fixing `VAST_TRACE` to handle `caf::message` is straight forward. On the other hand, introducing more complexity in the code base through a separate logging mechanism doesn't seem to be warranted in my eyes.

Reply:
> Wouldn't to_string(msg) work with VAST's logging macros?

Wouldn't that print `to_string(msg) = ...` in the logs? I miss the `CAF_ARG` functionality anyway, so I think we should talk about refactoring the logging parts sooner rather than later.

Reply:
Alright, let's get over with this and tackle the logging refactoring next.